pkg/controllers/user/monitoring/config_refresh_handler.go

```diff
@@ -0,0 +1,112 @@
+package monitoring
+
+import (
+	"encoding/json"
+	"reflect"
+	"sort"
+
+	"github.com/rancher/rancher/pkg/controllers/user/nslabels"
+	apiv1 "github.com/rancher/types/apis/core/v1"
+	rmonitoringv1 "github.com/rancher/types/apis/monitoring.coreos.com/v1"
+
+	monitoringv1 "github.com/coreos/prometheus-operator/pkg/client/monitoring/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	projectNSAnnotation = "project.cattle.io/namespaces"
+)
+
+type ConfigRefreshHandler struct {
+	prometheusClient rmonitoringv1.PrometheusInterface
+	prometheusLister rmonitoringv1.PrometheusLister
+	nsClient         apiv1.NamespaceInterface
```

Comment:
why did you chooose client over lister?

Reply:
clients are much more expense to make list and get calls, which is one way you are uisng them

Reply:
Updated